### PR TITLE
[HttpFoundation] fix session

### PR DESCRIPTION
**Pull request by ronfroy**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| License       | MIT

`getSession` must not return a callable when using `setSessionFactory`

### CODE REVIEW

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 1]
```diff
@@ -707,16 +707,17 @@ public function get($key, $default = null)
...
         }
 
-        if (null === $session) {
+        if (null === $this->session) {
             @trigger_error(sprintf('Calling "%s()" when no session has been set is deprecated since Symfony 4.1 and will throw an exception in 5.0. Use "hasSession()" instead.', __METHOD__), E_USER_DEPRECATED);
             // throw new \BadMethodCallException('Session has not been set');
         }
 
-        return $session;
+        return $this->session;
```
Comment (ronfroy) [inline src/Symfony/Component/HttpFoundation/Request.php]: must not return a callable when using `setSessionFactory`

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 2]
```diff
@@ -707,16 +707,17 @@ public function get($key, $default = null)
     public function getSession()
     {
         $session = $this->session;
-        if (!$session instanceof SessionInterface && null !== $session) {
+
+        if (is_callable($session)) {
```
Comment (ronfroy) [inline src/Symfony/Component/HttpFoundation/Request.php]: this is a more readable condition

**Comment by ronfroy** [comment]
@nicolas-grekas 

`setSessionFactory(function(){});` then

`getSession()` => return a callable on the first call

**Comment by nicolas-grekas** [comment]
Not quite:
```php
$r = new Request();
$r->setSessionFactory(function(){});
dump($r->getSession());
```
Leads to: `TypeError: Argument 1 passed to Symfony\Component\HttpFoundation\Request::setSession() must be an instance of Symfony\Component\HttpFoundation\Session\SessionInterface, null given`, which is legit since the factory doesn't respect the contract.

**Comment by ronfroy** [comment]
@nicolas-grekas the return of the Factory is also fixed if it is not a SessionInterface.

**Comment by ronfroy** [comment]
```$r = new Request();
$r->setSessionFactory(function(){ return new Session()});
$r->getSession(); => Session Interface```

```$r = new Request();
$r->setSessionFactory(function(){ });
$r->getSession(); => Throw exception```

**Comment by nicolas-grekas** [comment]
I'm sorry but I still fail to see what this fixes at all.
If you found a bug, please add a test case that fails without the patch, that'd end the discussion.
If it's a pure refactoring, then better not merge the PR as everything would be fine.

**Comment by ronfroy** [comment]
@nicolas-grekas sorry, yes we have to close it.

**Pull request closed without merging**